### PR TITLE
Add CRD feature and add KNP tests

### PR DESCRIPTION
**Pull request by doniacld**
## Motivation

Adding Kubernetes Network Policy tests coverage.
 
## Changes

- Add K8S extension client to retrieve CRD
- Add CRD feature and CNP enabled requirement
- Add Kubernetes Network Policies
- Add Kubernetes Network Policies tests scenarios

## List of scenarios added for Kubernetes Network Policy

| Scenario                                   | CNP                                                                                | KNP                                               | Comment                               |
|--------------------------------------------|------------------------------------------------------------------------------------|---------------------------------------------------|---------------------------------------|
| no-policies                                | -                                                                                  | -                                                 |                                       |
| no-policies-extra                          | -                                                                                  | -                                                 |                                       |
| allow-all-except-world                     | manifests/allow-all-except-world.yaml                                              | ❌                                                 | no entities in KNP                    |       
| client-ingress                             | manifests/client-ingress-from-client2.yaml                                         | manifests/client-ingress-from-client2-knp.yaml    |                                       |
| all-ingress-deny                           | manifests/deny-all-ingress.yaml                                                    | manifests/deny-all-ingress-knp.yaml               |                                       |
| all-egress-deny                            | manifests/deny-all-egress.yaml                                                     | manifests/deny-all-egress-knp.yaml                |                                       |
| all-entities-deny                          | manifests/deny-all-entities.yaml                                                   | ❌                                                 | no entities in KNP                    |
| cluster-entity                             | manifests/allow-cluster-entity.yaml                                                | ❌                                                 | no entities in KNP                    |
| host-entity                                | manifests/allow-host-entity.yaml                                                   | ❌                                                 | no entities in KNP                    |
| echo-ingress                               | manifests/echo-ingress-from-other-client.yaml                                      | manifests/echo-ingress-from-other-client-knp.yaml |                                       |
| client-ingress-icmp                        | manifests/echo-ingress-icmp.yaml                                                   | ❌                                                 | no icmp in KNP                        |
| client-egress                              | manifests/client-egress-to-echo.yaml                                               | manifests/client-egress-to-echo-knp.yaml          |                                       |
| client-egress-expression                   | manifests/client-egress-to-echo-expression.yaml                                    | manifests/client-egress-to-echo-expression-knp.yaml                                                 |  |
| client-egress-to-echo-service-account      | manifests/client-egress-to-echo-service-account.yaml                               | ❌                                                 | no Cilium identity labels in KNP      |
| to-entities-world                          | manifests/client-egress-to-entities-world.yaml                                     | ❌                                                 | no entities in KNP                    |
| to-cidr-1111                               | manifests/client-egress-to-cidr-1111.yaml                                          | manifests/client-egress-to-cidr-1111-knp.yaml                                                  | ipblock supported by Cilium for external IPs    |
| echo-ingress-from-other-client-deny        | manifests/echo-ingress-from-other-client-deny.yaml                                 | ❌                                                 | deny not supported by KNP             |
| client-ingress-from-other-client-icmp-deny | manifests/echo-ingress-icmp-deny.yaml                                              | ❌                                                 | no icmp in KNP                        |
| client-egress-to-echo-deny                 | manifests/client-egress-to-echo-deny.yaml                                          | ❌                                                 | deny not supported by KNP             |
| client-ingress-to-echo-named-port-deny     | manifests/client-egress-to-echo-named-port-deny.yaml                               | ❌                                                 | deny not supported by KNP             |
| client-egress-to-echo-expression-deny      | manifests/client-egress-to-echo-expression-deny.yaml                               | ❌                                                 | deny not supported by KNP             |
| client-egress-to-echo-service-account-deny | manifests/echo-ingress-from-other-client-deny.yaml                                 | ❌                                                 | no Cilium identity labels in KNP      |
| client-egress-to-cidr-deny                 | manifests/client-egress-to-cidr-1111-deny.yaml                                     | ❌                                                 | ideny not supported by KNP    |
| client-egress-to-cidr-deny-default         | manifests/client-egress-to-cidr-1111-deny.yaml                                     | ❌                                                 | deny not supported by KNP    |
| echo-ingress-l7                            | manifests/echo-ingress-l7-http.yaml                                                | ❌                                                 | L7 not supported by KNP               |
| echo-ingress-l7-named-port                 | manifests/echo-ingress-l7-http-named-port.yaml                                     | ❌                                                 | L7 not supported by KNP               |
| client-egress-l7-method                    | manifests/client-egress-only-dns.yaml, manifests/client-egress-l7-http-method.yaml | ❌                                                 | L7 not supported by KNP               |
| client-egress-l7                           | manifests/client-egress-l7-http.yaml                                               | ❌                                                 | L7 not supported by KNP               |
| client-egress-l7-named-port                | manifests/client-egress-l7-http-named-port.yaml                                    | ❌                                                 | L7 not supported by KNP               |
| client-egress-l7-tls-deny-without-headers  | manifests/client-egress-l7-tls.yaml                                                | ❌                                                 | L7 not supported by KNP               |
| client-egress-l7-tls-headers               | manifests/client-egress-l7-tls.yaml                                                | ❌                                                 | L7 not supported by KNP               |
| client-egress-l7-set-header                | manifests/client-egress-l7-http-matchheader-secret.yaml                            | ❌                                                 | L7 not supported by KNP               |
| dns-only                                   | manifests/client-egress-only-dns.yaml                                              | ❌                                                 | L7 (dns) not supported by KNP         |
| to-fqdns                                   | manifests/client-egress-to-fqdns-one-one-one-one.yaml                              | ❌                                                 | fqdns not supported by KNP            |

Fixes: https://github.com/cilium/cilium-cli/issues/1484

### CODE REVIEW

**Review by michi-covalent — APPROVED** [review]

File: connectivity/check/features.go [hunk 1]
```diff
@@ -473,6 +485,34 @@ func (ct *ConnectivityTest) ForceDisableFeature(feature Feature) {
 	ct.features[feature] = FeatureStatus{Enabled: false}
 }
 
+// isFeatureKNPEnabled checks if the Kubernetes Network Policy feature is enabled from the configuration.
+// Note that the flag appears in Cilium version 1.14, before that it was unable even thought KNPs were present.
```
Comment (joestringer) [inline connectivity/check/features.go:489]: Is there a PR or issue somewhere tracking this? I don't see the flag in the main branch of cilium/cilium. (I'd be completely fine with just replacing the internals of this function with `return true` then if we deem it useful to add a flag to cilium/cilium, follow up on that)

File: connectivity/suite.go
Comment (joestringer) [inline connectivity/suite.go:1]: The PR description states:

> ipblock not yet implemented by Cilium

This is not true, Cilium supports `ipBlock`. Whatever CIDR tests we have for CNP, the equivalent `ipBlock` statements in KNP should also work exactly the same way.

**Review by joestringer — CHANGES_REQUESTED** [review]
Nice work, tidy submission. I didn't look super closely at the details but it seems good overall :+1: 

Main blocker is that `ipBlock` policy is supported by Cilium so it should be possible to add those tests in as well.

File: connectivity/suite.go
Comment (doniacld) [reply connectivity/suite.go:1]: Fixed it by adding a new test case using CIDR on KNP and updated the PR description.

File: connectivity/check/features.go (hunk 1 above)
Comment (doniacld) [reply connectivity/check/features.go:489]: I am referring to the following flag in Cilium:
https://github.com/cilium/cilium/blob/87c8746e254d2b3ccdf8a42bb48367b6ecf2890f/pkg/defaults/defaults.go#L521
Maybe my comment is not clear... 🤨

File: connectivity/manifests/client-egress-to-cidr-1111-knp.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# This policy allows packets towards {{.ExternalIP}}, but not {{.ExternalOtherIP}}.
+# Both addresses are owned by CloudFlare/APNIC.
```
Comment (joestringer) [inline connectivity/manifests/client-egress-to-cidr-1111-knp.yaml:2]: nit, I'm guessing this was copied from somewhere but if the IPs are templated then we don't know if they're CloudFlare/APNIC addresses.
```suggestion
```

**Review by joestringer — APPROVED** [review]
I mainly looked at the network policy manifests, LGTM. Will leave the remaining review to codeowners.

File: connectivity/manifests/client-egress-to-cidr-1111-knp.yaml (hunk 2 above)
Comment (doniacld) [reply connectivity/manifests/client-egress-to-cidr-1111-knp.yaml]: Fixed!

**Comment by michi-covalent** [comment]
hey @sayboras could you review this? @youngnick and @tklauser might not be around.

**Review by sayboras — APPROVED** [review]
Nice work :heavy_check_mark:

**Pull request merged**